<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>nf_iap_readme</title>
  <style>
    html {
      line-height: 1.5;
      font-family: Georgia, serif;
      font-size: 20px;
      color: #1a1a1a;
      background-color: #fdfdfd;
    }
    body {
      margin: 0 auto;
      max-width: 36em;
      padding-left: 50px;
      padding-right: 50px;
      padding-top: 50px;
      padding-bottom: 50px;
      hyphens: auto;
      overflow-wrap: break-word;
      text-rendering: optimizeLegibility;
      font-kerning: normal;
    }
    @media (max-width: 600px) {
      body {
        font-size: 0.9em;
        padding: 1em;
      }
    }
    @media print {
      body {
        background-color: transparent;
        color: black;
        font-size: 12pt;
      }
      p, h2, h3 {
        orphans: 3;
        widows: 3;
      }
      h2, h3, h4 {
        page-break-after: avoid;
      }
    }
    p {
      margin: 1em 0;
    }
    a {
      color: blue;
    }
    a:visited {
      color: blue;
    }
    img {
      max-width: 100%;
    }
    h1, h2, h3, h4, h5, h6 {
      margin-top: 1.4em;
    }
    h5, h6 {
      font-size: 1em;
      font-style: italic;
    }
    h6 {
      font-weight: normal;
    }
    ol, ul {
      padding-left: 1.7em;
      margin-top: 1em;
    }
    li > ol, li > ul {
      margin-top: 0;
    }
    blockquote {
      margin: 1em 0 1em 1.7em;
      padding-left: 1em;
      border-left: 2px solid #e6e6e6;
      color: #606060;
    }
    code {
      font-family: Menlo, Monaco, 'Lucida Console', Consolas, monospace;
      font-size: 85%;
      margin: 0;
    }
    pre {
      margin: 1em 0;
      overflow: auto;
    }
    pre code {
      padding: 0;
      overflow: visible;
      overflow-wrap: normal;
    }
    .sourceCode {
     background-color: transparent;
     overflow: visible;
    }
    hr {
      background-color: #1a1a1a;
      border: none;
      height: 1px;
      margin: 1em 0;
    }
    table {
      margin: 1em 0;
      border-collapse: collapse;
      width: 100%;
      overflow-x: auto;
      display: block;
      font-variant-numeric: lining-nums tabular-nums;
    }
    table caption {
      margin-bottom: 0.75em;
    }
    tbody {
      margin-top: 0.5em;
      border-top: 1px solid #1a1a1a;
      border-bottom: 1px solid #1a1a1a;
    }
    th {
      border-top: 1px solid #1a1a1a;
      padding: 0.25em 0.5em 0.25em 0.5em;
    }
    td {
      padding: 0.125em 0.5em 0.25em 0.5em;
    }
    header {
      margin-bottom: 4em;
      text-align: center;
    }
    #TOC li {
      list-style: none;
    }
    #TOC a:not(:hover) {
      text-decoration: none;
    }
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<nav id="TOC" role="doc-toc">
<ul>
<li><a href="#nf-iap-pipeline">NF-IAP PIPELINE</a>
<ul>
<li><a href="#download-and-information">Download and information</a></li>
<li><a href="#output-description">Output description</a>
<ul>
<li><a href="#multiqc-information">MultiQC information</a></li>
</ul></li>
<li><a href="#software">Software</a></li>
<li><a href="#parameters">Parameters</a></li>
<li><a href="#resources">Resources</a></li>
<li><a href="#material-and-methods">Material and Methods</a></li>
</ul></li>
</ul>
</nav>
<h1 id="nf-iap-pipeline">NF-IAP PIPELINE</h1>
<p>NF-IAP is a Nextflow Illumina Analysis Pipeline meant for processing of DNA (WGS/WES) data.</p>
<p><a href="https://github.com/UMCUGenetics/NF-IAP">https://github.com/UMCUGenetics/NF-IAP</a></p>
<p><strong>The pipeline performs the following tasks.</strong></p>
<ul>
<li>Read QC (FastQC)</li>
<li>Mapping (BWA)</li>
<li>QC (GATK)</li>
<li>PCR duplicate detection (Sambamba MarkDup)</li>
<li>Variant calling and filtering (GATK)</li>
<li>Variant annotation (snpEff and GATK)</li>
<li>CNV calling (Control-FREEC)</li>
<li>SV calling (Manta)</li>
<li>QC report (MultiQC)</li>
</ul>
<h2 id="download-and-information">Download and information</h2>
<p>The pipeline can be downloaded from the link below:</p>
<p><a href="https://github.com/UMCUGenetics/NF-IAP">https://github.com/UMCUGenetics/NF-IAP</a></p>
<p>Follow the “Installing &amp; Setup” section to set up the pipeline for own use.</p>
<h2 id="output-description">Output description</h2>
<p>Descriptions of the contents of the resulting processed data folder.</p>
<ul>
<li><p><strong>.nextflow/</strong><br />
Nextflow cache and logging files, not of importance for the user but kept for completeness</p></li>
<li><p><strong>BAMS/</strong><br />
The resulting aligned data in .bam format with corresponding .bam.bai index files. These can be viewed in for example the IGV genome browser or used for further analysis</p></li>
<li><p><strong>CNV/</strong><br />
The Copy Number Variation results, when requested, for the corresponding tools in this pipeline</p>
<ul>
<li><strong>CNV/FREEC/</strong><br />
Control-FREEC CNV results</li>
</ul></li>
<li><p><strong>configs/</strong><br />
A copy various general configuration files used for the pipeline, containing parameters, resources, etc</p></li>
<li><p><strong>log/</strong><br />
Pipeline logs and nextflow report files (regarding resource usage, time taken, etc)</p></li>
<li><p><strong>QC/</strong><br />
Various QC logs and reports</p>
<ul>
<li><p><strong>QC/fastqc/</strong><br />
FastQC reports on the raw fastq files</p></li>
<li><p><strong>QC/multiple_metrics/</strong><br />
Collection of metrics as collected by CollectMultipleMetrics (Picard), such as: gc bias, base qualityscore distribution, etc. See <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037594031-CollectMultipleMetrics-Picard-">https://gatk.broadinstitute.org/hc/en-us/articles/360037594031-CollectMultipleMetrics-Picard-</a></p></li>
<li><p><strong>QC/summary/</strong><br />
MultiQC collects all information from the various QC modules as well as other tools (STAR, sambamba) and shows this in a summary overview in order to get a good and quick overview of the statistics of the various steps in the pipeline for this run</p></li>
<li><p><strong>QC/wgs_metrics/</strong><br />
Collection of metrics about coverage and performance of WGS experiments as collected by CollectWgsMetrics (Picard). See: <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037269351-CollectWgsMetrics-Picard-">https://gatk.broadinstitute.org/hc/en-us/articles/360037269351-CollectWgsMetrics-Picard-</a></p></li>
</ul></li>
<li><p><strong>SV/</strong><br />
The Structural Variants results, when requested, for the corresponding tools in this pipeline</p>
<ul>
<li><strong>SV/MANTA/</strong><br />
Manta SV results</li>
</ul></li>
<li><p><strong>VCFS/</strong><br />
The variants called in both VCF and GVCF format for the samples in this run</p>
<ul>
<li><strong>VCFS/VCF/</strong><br />
The variants called with GATK in VCF format, both filtered and annotated versions when requested<br />
</li>
<li><strong>VCFS/GVCF/</strong><br />
GVCF file for each sample</li>
</ul></li>
<li><p><strong>work/</strong><br />
The Nextflow working directory containing logs and script from the various cached steps. Kept for completeness and not of importance for the user</p></li>
</ul>
<h3 id="multiqc-information">MultiQC information</h3>
<p>MultiQC generates a summary report using the logs and reports generated by a selection of the various tools in the piplines.</p>
<h4 id="general-statistics">General Statistics</h4>
<p>The General Statistics provide an overview of the various modules, by hoovering over the column headers one can see from which specific tool this column was gathered. The Sample Name column lists both the actual sample (with info regarding the aligned bam file) as well as the various fastq files containing the raw reads for that sample (and the corresponding fastqc information).</p>
<h4 id="picard">Picard</h4>
<p>These metrics are generated using the GATK (picard) toolkit, see the relevant documentation for more indepth information on these modules:<br />
<a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037269351-CollectWgsMetrics-Picard-">CollectWgsMetrics (Picard)</a><br />
<a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036510492-CollectMultipleMetrics-Picard-">CollectMultipleMetrics (Picard)</a></p>
<h4 id="fastqc">FastQC</h4>
<p>Please refer to the FastQC documentation for the interpretation of the various modules:<br />
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p>
<h2 id="software">Software</h2>
<p>Tools used:</p>
<ul>
<li>Samtools value_samtools<br />
<a href="http://www.htslib.org/doc/samtools.html">http://www.htslib.org/doc/samtools.html</a><br />
</li>
<li>BWA v0.7.17<br />
<a href="http://bio-bwa.sourceforge.net/">http://bio-bwa.sourceforge.net/</a><br />
</li>
<li>Control-FREEC v11.5<br />
<a href="http://boevalab.inf.ethz.ch/FREEC/">http://boevalab.inf.ethz.ch/FREEC/</a><br />
</li>
<li>FastQC v0.11.5<br />
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a><br />
</li>
<li>GATK v4.1.3.0<br />
<a href="https://gatk.broadinstitute.org/hc/en-us">https://gatk.broadinstitute.org/hc/en-us</a><br />
</li>
<li>Manta v1.6.0<br />
<a href="https://github.com/Illumina/manta">https://github.com/Illumina/manta</a><br />
</li>
<li>MultiQC<br />
<a href="https://multiqc.info/">https://multiqc.info/</a><br />
</li>
<li>Sambamba v0.6.8<br />
<a href="https://lomereiter.github.io/sambamba/docs/sambamba-flagstat.html">https://lomereiter.github.io/sambamba/docs/sambamba-flagstat.html</a><br />
</li>
<li>snpEff v4.3.t<br />
<a href="http://pcingola.github.io/SnpEff/">http://pcingola.github.io/SnpEff/</a></li>
</ul>
<h2 id="parameters">Parameters</h2>
<p>When parameters are not specifically specified, the default parameters are used</p>
<pre><code>// Custom settings of tools.
params.freec_ploidy = 2
params.freec_window = 1000
params.freec_telocentromeric = 50000
params.freec_maxlevel = 4
  
params.bwa.optional = &#39;-M -c 100&#39;
params.bwaindex.optional = &#39;-a bwtsw&#39;
  
params.haplotypecaller.optional = &#39;-ERC GVCF&#39;      
params.collectmultiplemetrics.optional = &#39;--PROGRAM CollectAlignmentSummaryMetrics --PROGRAM CollectInsertSizeMetrics --PROGRAM QualityScoreDistribution --PROGRAM MeanQualityByCycle --PROGRAM CollectBaseDistributionByCycle --PROGRAM CollectGcBiasMetrics --PROGRAM CollectSequencingArtifactMetrics --PROGRAM CollectQualityYieldMetrics&#39;
params.markdup.optional = &#39;--overflow-list-size=2000000&#39;
  
params.snpefffilter.optional = &#39;GRCh37.75 -hgvs -lof -no-downstream -no-upstream -no-intergenic&#39;
params.snpsiftsbnsfp.optional = &#39;-f hg38_chr,hg38_pos,genename,Uniprot_acc,Uniprot_id,Uniprot_aapos,Interpro_domain,cds_strand,refcodon,SLR_test_statistic,codonpos,fold-degenerate,Ancestral_allele,Ensembl_geneid,Ensembl_transcriptid,aapos,aapos_SIFT,aapos_FATHMM,SIFT_score,SIFT_converted_rankscore,SIFT_pred,Polyphen2_HDIV_score,Polyphen2_HDIV_rankscore,Polyphen2_HDIV_pred,Polyphen2_HVAR_score,Polyphen2_HVAR_rankscore,Polyphen2_HVAR_pred,LRT_score,LRT_converted_rankscore,LRT_pred,MutationTaster_score,MutationTaster_converted_rankscore,MutationTaster_pred,MutationAssessor_score,MutationAssessor_rankscore,MutationAssessor_pred,FATHMM_score,FATHMM_rankscore,FATHMM_pred,MetaSVM_score,MetaSVM_rankscore,MetaSVM_pred,MetaLR_score,MetaLR_rankscore,MetaLR_pred,Reliability_index,VEST3_score,VEST3_rankscore,PROVEAN_score,PROVEAN_converted_rankscore,PROVEAN_pred,CADD_raw,CADD_raw_rankscore,CADD_phred,GERP++_NR,GERP++_RS,GERP++_RS_rankscore,phyloP46way_primate,phyloP46way_primate_rankscore,phyloP46way_placental,phyloP46way_placental_rankscore,phyloP100way_vertebrate,phyloP100way_vertebrate_rankscore,phastCons46way_primate,phastCons46way_primate_rankscore,phastCons46way_placental,phastCons46way_placental_rankscore,phastCons100way_vertebrate,phastCons100way_vertebrate_rankscore,SiPhy_29way_pi,SiPhy_29way_logOdds,SiPhy_29way_logOdds_rankscore,LRT_Omega,UniSNP_ids,1000Gp1_AC,1000Gp1_AF,1000Gp1_AFR_AC,1000Gp1_AFR_AF,1000Gp1_EUR_AC,1000Gp1_EUR_AF,1000Gp1_AMR_AC,1000Gp1_AMR_AF,1000Gp1_ASN_AC,1000Gp1_ASN_AF,ESP6500_AA_AF,ESP6500_EA_AF,ARIC5606_AA_AC,ARIC5606_AA_AF,ARIC5606_EA_AC,ARIC5606_EA_AF,ExAC_AC,ExAC_AF,ExAC_Adj_AC,ExAC_Adj_AF,ExAC_AFR_AC,ExAC_AFR_AF,ExAC_AMR_AC,ExAC_AMR_AF,ExAC_EAS_AC,ExAC_EAS_AF,ExAC_FIN_AC,ExAC_FIN_AF,ExAC_NFE_AC,ExAC_NFE_AF,ExAC_SAS_AC,ExAC_SAS_AF,clinvar_rs,clinvar_clnsig,clinvar_trait,COSMIC_ID,COSMIC_CNT&#39;
params.snpsiftannotate.optional = &#39;-tabix -name GoNLv5 -info AF,AN,AC&#39;</code></pre>
<h2 id="resources">Resources</h2>
<p>The following resources and pipeline/nextflow versions were used for this run:</p>
<ul>
<li><em>Pipeline Version:</em><br />
value_pipeline<br />
</li>
<li><em>Nextflow Version:</em><br />
value_nextflow<br />
</li>
<li><em>Genome:</em><br />
value_genome<br />
</li>
<li><em>Genome fasta:</em><br />
value_fasta</li>
</ul>
<h2 id="material-and-methods">Material and Methods</h2>
<p><strong>NF-IAP analysis</strong></p>
<p>Quality control on the sequence reads from the raw FASTQ files was done with FastQC (v0.11.5). Reads were aligned to the reference genome (value_genome) using BWA-mem (v0.7.17) after which they were sorted and index with Samtools (value_samtools). Duplicates were marked using Sambamba (v0.6.8). Followup QC on the mapped (bam) files was done using GATK (v4.1.3.0) CollectMultipleMetrics and CollectWGSMetrics. Variant calling is done using GATK best practises, including BaseQualityScoreRecalibration (BQSR), HaplotypeCaller, VariantSelection and VariantFiltration. Annotation on the resulting VCF file was done using GATK (v4.1.3.0) and/or snpEff (v4.3.t), depending on the organism/genome. Control-FREEC (v0.11.5) was used for CNV analysis, while Manta (v1.6.0) was used for SV analysis. Finally a summary report was created using MultiQC (v1.5).</p>
</body>
</html>
